engines/hypno/scene.cpp

```diff
@@ -0,0 +1,371 @@
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include <typeinfo>
+#include "common/events.h"
+
+#include "hypno/grammar.h"
+#include "hypno/hypno.h"
+
+namespace Hypno {
+
+extern int parse_mis(const char *);
+
+const static char *sceneVariables[] = {
+	"GS_NONE",
+	"GS_SCTEXT",
+	"GS_AMBIENT",
+	"GS_MUSIC",
+	"GS_VOLUME",
+	"GS_MOUSESPEED",
+	"GS_MOUSEON",
+	"GS_LEVELCOMPLETE",
+	"GS_LEVELWON",
+	"GS_CHEATS",
+	"GS_SWITCH0",
+	"GS_SWITCH1",
+	"GS_SWITCH2",
+	"GS_SWITCH3",
+	"GS_SWITCH4",
+	"GS_SWITCH5",
+	"GS_SWITCH6",
+	"GS_SWITCH7",
+	"GS_SWITCH8",
+	"GS_SWITCH9",
+	"GS_SWITCH10",
+	"GS_SWITCH11",
+	"GS_SWITCH12",
+	"GS_COMBATJSON",
+	"GS_COMBATLEVEL",
+	"GS_PUZZLELEVEL",
+	NULL};
+
+void HypnoEngine::parseScene(const Common::String &prefix, const Common::String &filename) {
+	debugC(1, kHypnoDebugParser, "Parsing %s", filename.c_str());
+	Common::String name = convertPath(filename);
+	if (!prefix.empty())
+		name = prefix + "/" + name;
+	Common::File test;
+	assert(test.open(name.c_str()));
+
+	const uint32 fileSize = test.size();
+	char *buf = (char *)malloc(fileSize + 1);
+	test.read(buf, fileSize);
+	test.close();
+	buf[fileSize] = '\0';
+	parse_mis(buf);
+	Level level;
+	level.scene.prefix = prefix;
+	level.scene.hots = *g_parsedHots;
+	_levels[name] = level;
+	free(buf);
+}
+
+void HypnoEngine::resetSceneState() {
+	uint32 i = 0;
+	while (sceneVariables[i]) {
+		_sceneState[sceneVariables[i]] = 0;
+		i++;
+	}
+}
+
+bool HypnoEngine::checkSceneCompleted() {
+	return _sceneState["GS_LEVELCOMPLETE"];
+}
+
+// Hotspots
+
+void HypnoEngine::clickedHotspot(Common::Point mousePos) {
+	Hotspots *hots = stack.back();
+	Hotspot selected;
+	bool found = false;
+	int rs = 100000000;
+	int cs = 0;
+	for (Hotspots::const_iterator it = hots->begin(); it != hots->end(); ++it) {
+		const Hotspot h = *it;
+		if (h.type != MakeHotspot)
+			continue;
+
+		cs = h.rect.width() * h.rect.height();
+		if (h.rect.contains(mousePos)) {
+			if (cs < rs) {
+				selected = h;
+				found = true;
+				rs = cs;
+			}
+		}
+	}
+	if (found) {
+		if (selected.smenu) {
+			assert(!selected.smenu->empty());
+			_nextHotsToAdd = selected.smenu;
+		}
+
+		for (Actions::const_iterator itt = selected.actions.begin(); itt != selected.actions.end(); ++itt) {
+			Action *action = *itt;
+			if (typeid(*action) == typeid(ChangeLevel))
```

Comment:
Is it possible to use the `switch()` statement here?

Reply:
I don't think it is possible, since the result of `typeid` is not a "real" integer:

> error: switch quantity not an integer
>   switch (typeid(*action))

If you know any way to convert this to a switch using `typeid`, I will be happy to change it.

Reply:
Removed `typeid` and refactored using a `switch`.